Add script that uses google sheets api to read newsletter events

Amyh11325 opened the pull request:
# Overview
- `event-generator-sheets.mjs` is a script similar to `recurring-event-generator.mjs` and `single-event-generator.mjs` but instead of reading data from a .cvs file downloaded from the ACM Newsletter spreadsheet, this script uses Google's API to read data directly from the spreadsheet.

# Testing
Run `node scripts/event-generator-sheets.mjs`
Check the output written to `output.json`
```
[{"title":"Studio Fall General Meeting","allDay":false,"start":1664503200000,"end":1664508600000,"location":"Haines Hall A18","committee":"studio","description":"Do you enjoy playing games? Want to learn how to make them? 🎮 💖 ACM Studio is hosting our Fall General Meeting this Thursday, September 29th @7:00PM in Haines A18. Here you can meet our wonderful board of officers and learn more about what we do here in Studio, from game jams to our Student Run Studios program to fun socials. If you have any interest in developing games from start to finish, we're the committee for you! Whether it's programming, art, narrative, sound design, UI/UX, or any other aspect, you are welcome to join, regardless of skill level 🎵 🎨 💻 !! Come check us out, and let's make some games together! See you there 💖","links":[{"text":"Facebook Event","href":"https://www.instagram.com/p/CivIm1wJfyr/","ext":true}]},{"title":"Ludum Dare 51 Kickoff","allDay":false,"start":1664587800000,"end":1664595000000,"location":"Engineering VI 289","committee":"studio","description":"Ready to open the school year with a bang🤯 ? The biannual game jam, Ludum Dare is having its 51st jam this Friday, September 30th. The 🕐72 hour countdown🕐  begins once the jam's theme is revealed at 3:00 pm. Studio will be hosting a kickoff in E6 289 at 6:30 pm on Friday, so come join us and meet your team 💖 and get on the game jam grind! Don't worry if you missed the initial sign ups! You can sign up for the waitlist if you're interested in looking for a team through Studio: https://forms.gle/rNESbkHmRUXE3sMeA. Get ready to brainstorm, create, and realize a full game in just 72 hours!","links":[{"text":"Facebook Event","href":"https://www.instagram.com/p/Cissl1WPwkF/","ext":true}]},{"title":"Cyber Fall GM","allDay":false,
...
```
Output should either include the event or print an error for the event

# Notes
- Originally wrote the script in typescript, but I was having trouble compiling it in this repo, so I switched to javascript
- Copied the parsing methods from `recurring-event-generator.mjs` but I don't think it works with the way the spreadsheet is currently formatted? (eg: STARTING Friday, November 13 (Recurring every Friday until Week 9))

matthewcn56 commented:
Nice work Amy, I can see that it works when I run the script for output.json! However, I think it would be **really great** if we could utilize Next.js ISR to have this run automatically and automatically updated! Do you think we could do what we did similar to opensource and anti-resume of having the events page utilize data that's pulled directly from the sheets and updated every day with the **getStaticProps** function and have it revalidate every day?

Amyh11325 commented:
> Nice work Amy, I can see that it works when I run the script for output.json! However, I think it would be **really great** if we could utilize Next.js ISR to have this run automatically and automatically updated! Do you think we could do what we did similar to opensource and anti-resume of having the events page utilize data that's pulled directly from the sheets and updated every day with the **getStaticProps** function and have it revalidate every day?

I can give it a shot! One concern I have though is that some of the events in the sheet aren't always formatted correctly in a way the script can read, so we might still want someone looking through the script logs every week? Right now the script just prints an error and skips the event, if it can't read it.

advaithg commented:
I think the event formatting issue is something I saw at least one time every week, and it's definitely ideal to have someone actually check the sheet logs! We can also start pushing for cprez to not make some of those errors when adding stuff onto the sheet!

advaithg commented:
For single events, this looks good, but we should be modifying the script itself to not run the recurring event generator! Ideally this should be a separate PR for single events, but in this case I think it's fine to only run the `getEventsOfWeek()` function for any week we want to update events for.
However the downside is this script is still going to have the same errors as the old one did (with links in description / date time formatting not being accurate / zoom locations) that remain to be manually resolved (except you'd have to do it in the actual google sheet this time) so it's essentially the same workflow barring the having to download CSV files part.
I'll leave it to @matthewcn56 to decide exactly what to do with the script but code wise it looks good for single events! :)

matthewcn56 commented on scripts/event-generator-sheets.mjs:
```diff
@@ -0,0 +1,169 @@
+        end: null,
+        committee: getCssStringFromCommittee(row[0]),
+        location: row[5],
+        description: row[6],
```
We can make this a lot safer by using [nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) which will check to see if the value on the left exists. If not, it supplies the value on the right, like so:
```js
description: row[6] ?? ''
```
Up to you if you would rather skip an event if no description is provided or if you want to display it with an empty description. 
(fyi the site crashed when description was undefined oop)

matthewcn56 commented on pages/events.js:
```diff
@@ -93,4 +93,18 @@ function Events() {
+			events: events,
+		},
+	};
+};
```
don't forget the revalidate prop for [ISR](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration)! (maybe make it update every hour or every day?)

matthewcn56 reviewed (changes requested):
Great work Amy! The events page looks really nice being populated now, but i just left a couple comments about how we can speed up performance by parallelizing some async functions with `Promise.all` and making the `getRecurringEventsFromWeek` function synchronous by taking out the async call. Also don't forget ISR!

Amyh11325 commented:
> Great work Amy! The events page looks really nice being populated now, but i just left a couple comments about how we can speed up performance by parallelizing some async functions with `Promise.all` and making the `getRecurringEventsFromWeek` function synchronous by taking out the async call. Also don't forget ISR!

Addressed your comments. Thanks for the suggestions! These are really great improvements and catches! Lmk if there's anything else.

matthewcn56 reviewed (approved):
Great work! I can tell immediately that it's a lot faster and catching errors is great. Now we just have to figure out how to wrap links and line breaks, now our events page has stuff in it :)
